File: src/main/java/at/searles/parsing/Recognizable.java
package at.searles.parsing;

/**
 * Everything that can be recognized
 */
public interface Recognizable {

    boolean recognize(ParserStream stream);

    interface Then extends Recognizable {
        Recognizable left();

        Recognizable right();

        @Override
        default boolean recognize(ParserStream stream) {
            long offset = stream.getOffset();

            long preStart = stream.getStart();
            long preEnd = stream.getEnd();

            if (!left().recognize(stream)) {
                return false;
            }

            long start = stream.getStart();

            if (!right().recognize(stream)) {
                if(stream.getOffset() != offset) {
                    throwIfNoBacktrack(stream);
                    stream.backtrackToOffset(offset);
                }

                stream.setStart(preStart);
                stream.setEnd(preEnd);
                return false;
            }

            stream.setStart(start);
            return true;
        }

        default void throwIfNoBacktrack(ParserStream stream) {
            if(!allowParserBacktrack()) {
                throw new ParserLookaheadException(this, stream);
            }
        }

        boolean allowParserBacktrack();

        default String createString() {
            String leftString = left().toString();
            String rightString = right().toString();

            if (left() instanceof Or) {
                leftString = "(" + leftString + ")";
            }

            if (right() instanceof Or || right() instanceof Then) {
                rightString = "(" + rightString + ")";
            }

            return leftString + " " + rightString;
        }
    }

    interface Or extends Recognizable {
        Recognizable first();

        Recognizable second();

        @Override
        default boolean recognize(ParserStream stream) {
            return first().recognize(stream) || second().recognize(stream);
        }

        default String createString() {
            String firstString = first().toString();
            String secondString = second().toString();

            if (second() instanceof Or) {
                secondString = "(" + secondString + ")";
            }

            return firstString + " | " + secondString;
        }
    }

    interface Opt extends Recognizable {
        Recognizable parent();

        @Override
        default boolean recognize(ParserStream stream) {
            stream.setStart(stream.getEnd());
            parent().recognize(stream);

            return true;
        }

        default String createString() {
            String parentString = parent().toString();

            if (parent() instanceof Or || parent() instanceof Then) {
                parentString = "(" + parentString + ")";
            }

            return parentString + "?";
        }
    }

    interface Rep extends Recognizable {
        Recognizable parent();

        @Override
        default boolean recognize(ParserStream stream) {
            long start = stream.getStart();

            while (parent().recognize(stream)) {
                // do nothing. Everything is done in 'recognize'
                stream.setStart(start);
            }

            return true;
        }

        default String createString() {
            String parentString = parent().toString();

            if (parent() instanceof Or || parent() instanceof Then) {
                parentString = "(" + parentString + ")";
            }

            return parentString + "*";
        }
    }
}

File: src/main/java/at/searles/parsing/Recognizable.kt
package at.searles.parsing

/**
 * Everything that can be recognized
 */
interface Recognizable {

    fun recognize(stream: ParserStream): Boolean

    interface Then : Recognizable {

        val left: Recognizable
        val right: Recognizable

        override fun recognize(stream: ParserStream): Boolean {
            val offset = stream.offset
            val preStart = stream.start
            val preEnd = stream.end

            if (!left.recognize(stream)) {
                return false
            }

            val start = stream.start

            if (!right.recognize(stream)) {
                if (stream.offset != offset) {
                    stream.requestBacktrackToOffset(this, offset)
                }

                stream.start = preStart
                stream.end = preEnd

                return false
            }

            stream.start = start
            return true
        }

        fun createString(): String {
            val leftString = when(left) {
                is Or -> "($left)"
                else -> "$left"
            }

            val rightString = when(right) {
                is Or -> "($right)"
                is Then -> "($right)"
                else -> "$left"
            }

            return "$leftString $rightString"
        }
    }

    interface Or : Recognizable {

        val choice0: Recognizable
        val choice1: Recognizable

        override fun recognize(stream: ParserStream): Boolean {
            return choice0.recognize(stream) || choice1.recognize(stream)
        }

        fun createString(): String {
            return if (choice1 is Or) {
                "$choice0 | ($choice1)"
            } else {
                "$choice0 | $choice1"
            }
        }
    }

    interface Opt : Recognizable {

        val parent: Recognizable

        override fun recognize(stream: ParserStream): Boolean {
            stream.start = stream.end
            parent.recognize(stream)
            return true
        }

        fun createString(): String {
            return if (parent is Or || parent is Then) {
                 "($parent)?"
            } else {
                "$parent?"
            }
        }
    }

    interface Rep : Recognizable {

        val parent: Recognizable

        override fun recognize(stream: ParserStream): Boolean {
            val start = stream.start

            while (parent.recognize(stream)) {
                stream.start = start
            }

            return true
        }

        fun createString(): String {
            return if (parent is Or || parent is Then) {
                "($parent)*"
            } else {
                "$parent*"
            }
        }
    }
}